pkg/webhookadmission/webhook.go

```diff
@@ -0,0 +1,268 @@
+package webhookadmission
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/libopenstorage/stork/drivers/volume"
+	"github.com/portworx/sched-ops/k8s"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/api/admission/v1beta1"
+	appv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/record"
+)
+
+const (
+	mutateWebHook            = "/mutate"
+	validateWebHook          = "/validate"
+	appSchedPrefix           = "/spec/template"
+	podSpecSchedPath         = "/spec/schedulerName"
+	storkScheduler           = "stork"
+	storkAdmissionController = "stork-webhooks-cfg"
+	secretName               = "servercert-secret"
+	privKey                  = "privKey"
+	privCert                 = "privCert"
+)
+
+// Controller for admission mutating webhook to initialise resources
+// with stork as scheduler, if given resources are using driver supported
+// by stork
+type Controller struct {
+	Recorder record.EventRecorder
+	Driver   volume.Driver
+	server   *http.Server
+	lock     sync.Mutex
+	started  bool
+}
+
+// Serve method for webhook server
+func (c *Controller) serveHTTP(w http.ResponseWriter, req *http.Request) {
+	if strings.Contains(req.URL.Path, mutateWebHook) {
+		c.processMutateRequest(w, req)
+	} else {
+		http.Error(w, "Unsupported request", http.StatusNotFound)
+	}
+}
+
+func (c *Controller) processMutateRequest(w http.ResponseWriter, req *http.Request) {
+	var admissionResponse *v1beta1.AdmissionResponse
+	var err error
+	var schedPath string
+	admissionReview := v1beta1.AdmissionReview{}
+	isStorkResource := false
+
+	decoder := json.NewDecoder(req.Body)
+	defer func() {
+		if err := req.Body.Close(); err != nil {
+			log.Warnf("Error closing decoder")
+		}
+	}()
+	if err := decoder.Decode(&admissionReview); err != nil {
+		log.Errorf("Error decoding admission review request: %v", err)
+		c.Recorder.Event(&admissionReview, v1.EventTypeWarning, "invalid admission review request", err.Error())
+		http.Error(w, "Decode error", http.StatusBadRequest)
+		return
+	}
+
+	// TODO: This log does not get reflected in stork pods, check other logger
+	arReq := admissionReview.Request
+	switch arReq.Kind.Kind {
+	case "StatefulSet":
+		var ss appv1.StatefulSet
+		if err = json.Unmarshal(arReq.Object.Raw, &ss); err != nil {
+			log.Errorf("Could not unmarshal admission review object: %v", err)
+			c.Recorder.Event(&admissionReview, v1.EventTypeWarning, "could not unmarshal ar object", err.Error())
```

Comment:
Don't need to do as part of this PR, but events to the admission review won't really be visible anywhere. We should either add it to the deployment/sts/pod or if we don't have that info then add it to the webhook config object

Reply:
will do